src/main/java/com/yahoo/bullet/storm/JoinBolt.java

```diff
@@ -108,142 +109,235 @@ public void execute(Tuple tuple) {
 
     @Override
     public void declareOutputFields(OutputFieldsDeclarer declarer) {
-        declarer.declare(new Fields(TopologyConstants.ID_FIELD, JOIN_FIELD, TopologyConstants.METADATA_FIELD));
+        // This is where the data for each queries go
+        declarer.declareStream(RESULT_STREAM, new Fields(ID_FIELD, RESULT_FIELD, METADATA_FIELD));
+        // This is where the metadata that is used for feedback is sent
+        declarer.declareStream(FEEDBACK_STREAM, new Fields(ID_FIELD, METADATA_FIELD));
     }
 
-    @Override
-    protected AggregationQuery createQuery(Tuple queryTuple) {
-        String queryString = queryTuple.getString(TopologyConstants.QUERY_POSITION);
+    private void onTick() {
+        // Force emit all the done queries or closed queries that are being rotated out.
+        Map<String, Querier> forceDone = bufferedQueries.rotate();
+        forceDone.entrySet().forEach(this::emitFinished);
+        // The active queries count is not updated for these since these queries are not in any map, so do it here
+        updateCount(activeQueriesCount, -forceDone.size());
+
+        // Close all the buffered windows and re-add them back to queries
+        Map<String, Querier> forceClosed = bufferedWindows.rotate();
+        forceClosed.entrySet().forEach(this::emitBufferedWindow);
+
+        // Categorize all in queries in non-partition mode and do the roll over or emit as necessary
+        handleCategorizedQueries(new QueryCategorizer().categorize(queries, false));
+    }
+
+    private void onQuery(Tuple tuple) {
+        String id = tuple.getString(TopologyConstants.ID_POSITION);
+        String query = tuple.getString(TopologyConstants.QUERY_POSITION);
+        Metadata metadata = (Metadata) tuple.getValue(TopologyConstants.QUERY_METADATA_POSITION);
+
+        Querier querier;
         try {
-            return new AggregationQuery(queryString, configuration);
-        } catch (JsonParseException jpe) {
-            emitError(queryTuple, Arrays.asList(Error.makeError(jpe, queryString)));
-        } catch (ParsingException pe) {
-            emitError(queryTuple, pe.getErrors());
+            querier = createQuerier(id, query, config);
+            Optional<List<BulletError>> optionalErrors = querier.initialize();
+            if (!optionalErrors.isPresent()) {
+                setupQuery(id, query, metadata, querier);
+                return;
+            }
+            emitErrorsAsResult(id, metadata, optionalErrors.get());
         } catch (RuntimeException re) {
-            log.error("Unhandled exception.", re);
-            emitError(queryTuple, Arrays.asList(Error.makeError(re, queryString)));
+            // Includes JSONParseException
```

Comment:
Is JsonParseException is a RuntimeException?

Reply:
Yup